### PR TITLE
[9/N] Fixes clang-tidy warnings in c10/util/*.h

### PR DESCRIPTION
**Pull request by cyyever**
Continued work to clean headers in c10/util.

### CODE REVIEW

File: c10/util/LeftRight.h [hunk 1]
```diff
@@ -204,14 +203,17 @@ class RWSafeLeftRightWrapper final {
...
+        [&readFunc](T const& data) { return std::forward<F>(readFunc)(data); });
   }
 
   template <typename F>
-  auto write(F&& writeFunc) -> typename c10::invoke_result_t<F, T&> {
-    return data_.withLock([&writeFunc](T& data) { return writeFunc(data); });
+  // NOLINTNEXTLINE(cppcoreguidelines-missing-std-forward)
+  auto write(F&& writeFunc) {
+    return data_.withLock(
+        [&writeFunc](T& data) { return std::forward<F>(writeFunc)(data); });
```
Comment (Skylion007) [inline c10/util/LeftRight.h:216]: Wait, this isn't quite right... you are forwarding a reference, not a universal reference. If you want to do this right, it needs to be a move capture in the `[]. I think nearly all the std::forward fixes in this file seems like false positives though.

File: c10/util/Synchronized.h [hunk 2]
```diff
@@ -28,7 +26,7 @@ class Synchronized final {
  public:
   Synchronized() = default;
   Synchronized(T const& data) : data_(data) {}
-  Synchronized(T&& data) : data_(data) {}
+  Synchronized(T&& data) : data_(std::move(data)) {}
```
Comment (Skylion007) [inline c10/util/Synchronized.h:29]: Shouldn't this be a perfect forward here?

**Review by Skylion007 — CHANGES_REQUESTED** [review]
I had a few concerns.

File: c10/util/Synchronized.h (hunk 2 above)
Comment (cyyever) [reply c10/util/Synchronized.h:29]: We can use perfect forward here and remove the above constructor.

File: c10/util/LeftRight.h (hunk 1 above)
Comment (cyyever) [reply c10/util/LeftRight.h:216]: Lambda capture and perfect forward is somewhat confusing. But I think this fix is correct as you can see from https://stackoverflow.com/questions/21238463/capture-by-universal-reference

File: c10/util/Synchronized.h (hunk 2 above)
Comment (cyyever) [reply c10/util/Synchronized.h:29]: @Skylion007  No, this is a move constructor, not universal reference. The fix is correct.

File: c10/util/LeftRight.h (hunk 1 above)
Comment (cyyever) [reply c10/util/LeftRight.h:216]: @Skylion007 I think the fix is correct. The function was immediately called with lock, so there is no danger of dangling reference. And https://stackoverflow.com/questions/21238463/capture-by-universal-reference explains well the syntax here.

Comment (Skylion007) [reply c10/util/LeftRight.h:216]: I am more worried about `[&writeFunc]` the writeFunc in the lambda will always be an lvalue right? so the std::forward is safe, but a noop.

**Comment by cyyever** [comment]
@Skylion007 perfect forwarding in lambda may be hard to understand, then we can revert them?

**Comment by Skylion007** [comment]
@cyyever take a look at this issue to see what I mean about the lambdas: https://github.com/llvm/llvm-project/issues/75891

**Comment by Skylion007** [comment]
Huh, seems this is actually a valid way to perfectly forward into a lambda... https://stackoverflow.com/questions/26831382/capturing-perfectly-forwarded-variable-in-lambda

**Comment by cyyever** [comment]
@Skylion007 In the case that [var=std::forward], is var a dependent variable so that we should std::move(var) in the lambda body? Lambda capture syntax is somewhat confusing here...

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**